shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm

```diff
@@ -357,81 +443,196 @@
-      canvas->drawPicture(picture_recorders_[view_id]->finishRecordingAsPicture());
-      canvas->flush();
-      did_submit &= frame->Submit();
+  // Resolve all pending GPU operations before allocating a new surface.
+  background_canvas->flush();
+  // Clipping the background canvas before drawing the picture recorders requires to
+  // save and restore the clip context.
+  SkAutoCanvasRestore save(background_canvas, /*doSave=*/true);
+  // Maps a platform view id to a vector of `FlutterPlatformViewLayer`.
+  LayersMap platform_view_layers;
+
+  auto did_submit = true;
+  auto num_platform_views = composition_order_.size();
+
+  for (size_t i = 0; i < num_platform_views; i++) {
+    int64_t platform_view_id = composition_order_[i];
+    sk_sp<RTree> rtree = platform_view_rtrees_[platform_view_id];
+    sk_sp<SkPicture> picture = picture_recorders_[platform_view_id]->finishRecordingAsPicture();
+
+    // Check if the current picture contains overlays that intersect with the
+    // current platform view or any of the previous platform views.
+    for (size_t j = i + 1; j > 0; j--) {
+      int64_t current_platform_view_id = composition_order_[j - 1];
+      SkRect platform_view_rect = GetPlatformViewRect(current_platform_view_id);
+      std::list<SkRect> intersection_rects =
+          rtree->searchNonOverlappingDrawnRects(platform_view_rect);
+      auto allocation_size = intersection_rects.size();
+
+      // For testing purposes, the overlay id is used to find the overlay view.
+      // This is the index of the layer for the current platform view.
+      auto overlay_id = platform_view_layers[current_platform_view_id].size();
+
+      // If the max number of allocations per platform view is exceeded,
+      // then join all the rects into a single one.
+      //
+      // TODO(egarciad): Consider making this configurable.
+      // https://github.com/flutter/flutter/issues/52510
+      if (allocation_size > kMaxLayerAllocations) {
+        SkRect joined_rect;
+        for (const SkRect& rect : intersection_rects) {
+          joined_rect.join(rect);
+        }
+        // Replace the rects in the intersection rects list for a single rect that is
+        // the union of all the rects in the list.
+        intersection_rects.clear();
+        intersection_rects.push_back(joined_rect);
+      }
+      for (SkRect& joined_rect : intersection_rects) {
+        // Get the intersection rect between the current rect
+        // and the platform view rect.
+        joined_rect.intersect(platform_view_rect);
+        // Subpixels in the platform may not align with the canvas subpixels.
+        // To workaround it, round the floating point bounds and make the rect slighly larger.
+        // For example, {0.3, 0.5, 3.1, 4.7} becomes {0, 0, 4, 5}.
+        joined_rect.setLTRB(std::nearbyint(joined_rect.left()), std::nearbyint(joined_rect.top()),
+                            std::ceil(joined_rect.right()), std::ceil(joined_rect.bottom()));
```

Comment:
Assuming that the origin is top left I believe this is what you want:
```
joined_rect.setLTRB(ceil(joined_rect.left()), ceil(joined_rect.top()),
                   floor(joined_rect.right()), floor(joined_rect.bottom()));
```

You don't want to round because it's non-deterministic.  We need to make a decision, should we expand the area, or shrink it?

The code i wrote is guaranteeing that we are trimming fractional pixels to create a rect with the smallest surface area.  The reason I believe we want that is that way we can guarantee that we have content that will be contained by it.  One way rounding could be bad is if a mask and a view go to the right edge of 10.5, if the masks right edge is rounded to 11 it is exceeding the bounds of his view which may be benign or erroneous depending on the implementation of the mask.

You have a better understanding of the code, i'm not sure if you want to shrink the surface area, but I'm pretty sure you don't want rounding, you want ceil or floor.

Reply:
> The code i wrote is guaranteeing that we are trimming fractional pixels to create a rect with the smallest surface area.

The rect should cover a bigger surface, not a smaller one. Otherwise, a row or column could end up missing pixels.

Reply:
Ok, that sounds good to me.  Just remove the rounding to guarantee the surface area gets bigger then, right?

Reply:
(std::floor(joined_rect.left()), std::floor(joined_rect.top()), std::ceil(joined_rect.right()), std::ceil(joined_rect.bottom())) should produce a bigger surface area. Is there any concern with floor for top, left and ceil for right, bottom?

Reply:
`(std::floor(joined_rect.left()), std::floor(joined_rect.top()), std::ceil(joined_rect.right()), std::ceil(joined_rect.bottom()))` should produce a bigger surface area. Is there any concern with floor for top, left and ceil for right, bottom?